app/org/maproulette/models/dal/VirtualProjectDAL.scala

```diff
@@ -43,13 +43,27 @@ class VirtualProjectDAL @Inject()(override val db: Database,
       case Some(p) => if (!p.isVirtual.getOrElse(false)) {
         throw new InvalidException(s"Project must be a virtual project to add a challenge.")
       }
-      case None => throw new NotFoundException(s"No challenge with id $challengeId found.")
+      case None => throw new NotFoundException(s"No project found with id $projectId found.")
     }
+
+
     this.withMRTransaction { implicit c =>
-      val query =
-        s"""INSERT INTO virtual_project_challenges (project_id, challenge_id)
-                      VALUES ($projectId, $challengeId)"""
-      SQL(query).execute()
+      val existing =
+        SQL(
+        """SELECT id FROM virtual_project_challenges
+           WHERE project_id = {projectId} AND challenge_id = {challengeId} limit 1
+        """)
+          .on('projectId -> projectId, 'challengeId -> challengeId)
+          .as(SqlParser.long("id").singleOpt)
+
+      existing match {
+        case Some(e) => // already in database, do not add again
+        case _ =>
+          val query =
+            s"""INSERT INTO virtual_project_challenges (project_id, challenge_id)
```

Comment:
So you don't have to do the check first. You can instead use a try catch around this insert and then just ignore any unique key violations. That way you only need to make a single call to the database instead of 2 every time. So something like:
```
try {
    ...
} catch {
    case e:PSQLException if (e.getSQLState == "23505") => //ignore
    case _ => // throw some exception here
}
```